### PR TITLE
adding intercept inbound channel

### PR DESCRIPTION
**Pull request by niteshbalusu11**
Added a new command called `intercept-inbound-channel`. 

added 2 new files: intercept_inbound_channel and generate_address_from_pubkey

installed a new package called bip32 used for address generation in bitcoinjs

The challenge is generating a new address on the fly everytime and where to store the derivation index to increment it the next time. Right now I hardcoded it to zero.

### CODE REVIEW

**Comment by alexbosworth** [comment]
Could this just be part of inbound-channel-rules?

**Comment by niteshbalusu11** [comment]
> Could this just be part of inbound-channel-rules?

I tried doing that, looks like inbound channel rules expects at least one rule, maybe that needs to change?

**Comment by alexbosworth** [comment]
OK, yeah maybe they can be merged I can take a look

also, does this always just use the index 0 address?

**Comment by alexbosworth** [comment]
There is an issue with the formatting in the `bos` file, it edited a lot of lines there

File: network/intercept_inbound_channel.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+          },
+        ],
+
+        //intercept open channel requests and generate addresses on the fly.
+        interceptOpenRequests: [
+          "validate",
+          "getFile",
+          "getAddress",
+          ({ getAddress, getFile }, cbk) => {
+            const sub = subscribeToOpenRequests({ lnd: args.lnd });
```
Comment (alexbosworth) [inline network/intercept_inbound_channel.js:73]: should probably listen for errors, what if this subscription has an issue

File: chain/generate_address_from_pubkey.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  if (network === "testnet") {
+    network = bjs.networks.testnet;
+  }
+  if (network === "regtest") {
+    network = bjs.networks.regtest;
+  }
+
+  const xpub = anypubToXpub(masterPubKey);
+
+  //generate an address, using p2sh wrapper over p2wpkh to support legacy wallets
```
Comment (alexbosworth) [inline chain/generate_address_from_pubkey.js:21]: is there some way to tell which type of master public key it is and then generate the correct address type for that key?

**Review by alexbosworth — CHANGES_REQUESTED** [review]

File: network/intercept_inbound_channel.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            return cbk([400, "ExpectedLoggerObjectToInterceptInboundChannel"]);
+          }
+          return cbk();
+        },
+
+        //read the master pubkey and network from file
+        getFile: [
+          "validate",
+          ({}, cbk) => {
+            const path = join(...[homedir(), home, pubkeyFileName]);
```
[the remark this answers is not in the document]
Comment (niteshbalusu11) [reply network/intercept_inbound_channel.js:32]: > There is an issue with the formatting in the `bos` file, it edited a lot of lines there

uhh, looks like prettier formatter just reformatted the file. Will try to fix it.

File: network/intercept_inbound_channel.js (hunk 1 above)
Comment (niteshbalusu11) [reply network/intercept_inbound_channel.js:73]: > should probably listen for errors, what if this subscription has an issue

okay, I can do that.

File: chain/generate_address_from_pubkey.js (hunk 2 above)
Comment (alexbosworth) [reply chain/generate_address_from_pubkey.js:21]: ok so if the xpub can't tell you what type of address it wants, how about having that as a parameter so you could have native segwit or nested segwit

**Comment by alexbosworth** [comment]
> > OK, yeah maybe they can be merged I can take a look
> > also, does this always just use the index 0 address?
> 
> This is part I need help with, index 0 and 1 works, any number above that, the address generates and gets tied to channel open but when you close it funds don't come back to it, so I am have to check how we can generate a unique address everytime. Also, where should we store the index value to just increment from the last value?

It doesn't make sense to me that higher index addresses wouldn't work

I can think about a way to do address incrementing, but if a single address is sufficient why not just ask for the address as an argument instead of the xpub?

**Comment by niteshbalusu11** [comment]
> > > OK, yeah maybe they can be merged I can take a look
> > > also, does this always just use the index 0 address?
> > 
> > 
> > This is part I need help with, index 0 and 1 works, any number above that, the address generates and gets tied to channel open but when you close it funds don't come back to it, so I am have to check how we can generate a unique address everytime. Also, where should we store the index value to just increment from the last value?
> 
> It doesn't make sense to me that higher index addresses wouldn't work
> 
> I can think about a way to do address incrementing, but if a single address is sufficient why not just ask for the address as an argument instead of the xpub?

The idea behing asking for an xpub is to follow the good practice of not reusing addresses, if we want to go the address route, that would make things so much easier.
The higher index part, I am scratching my head about that too on why it isn't working.

**Comment by alexbosworth** [comment]
> > > > OK, yeah maybe they can be merged I can take a look
> > > > also, does this always just use the index 0 address?
> > > 
> > > 
> > > This is part I need help with, index 0 and 1 works, any number above that, the address generates and gets tied to channel open but when you close it funds don't come back to it, so I am have to check how we can generate a unique address everytime. Also, where should we store the index value to just increment from the last value?
> > 
> > 
> > It doesn't make sense to me that higher index addresses wouldn't work
> > I can think about a way to do address incrementing, but if a single address is sufficient why not just ask for the address as an argument instead of the xpub?
> 
> The idea behing asking for an xpub is to follow the good practice of not reusing addresses, if we want to go the address route, that would make things so much easier. The higher index part, I am scratching my head about that too on why it isn't working.

if you ask for an xpub but reuse the same index, it's reusing addresses

maybe start with address if that is good enough? or the user could attach a bucket of addresses like --address --address --address?

**Comment by niteshbalusu11** [comment]
Ok, will start with an address. One address per saved node?

**Comment by niteshbalusu11** [comment]
> > Ok, will start with an address. One address per saved node?
> 
> if it's an address I think you could just pass it as part of the command, so like `--coop-close-address <address>`

Cool, will work on it.

**Comment by niteshbalusu11** [comment]
@alexbosworth can you check now?

**Comment by alexbosworth** [comment]
Looks cool I will try it out

**Review by alexbosworth — APPROVED** [review]

**Pull request merged**